### DISCUSSION TITLE
JaVers, JTA, and Spring Boot 3

### DISCUSSION THREAD

**Discussion by ghost** [javers/javers#1332] [Q&A]
Hi all

I am having an issue getting the latest versions of JaVers and JTA (Atomikos implementation) to work together on a Spring Boot 3 project. After some debugging, I have managed to narrow the issue down to the fact that JaVers is receiving an already closed connection from the entity manager via the `ConnectionProvider.getConnection()` method, which is then throwing a SQL exception when it tries to execute its SQL statements.

When I remove the JTA dependencies and `@Transactional` annotations for the project and revert to native JPA everything works fine, however I specifically require transactions in my project, so this is not a feasible solution.

Looking at the Hibernate documentation and suggestions, it is my understanding that the recommended way to perform work on a session is to pass it in to the `doWork()` and `doReturningWork()` methods rather than using the connection directly the way that JaVers currently does. I tested this by creating my own `ConnectionProvider` implementation that then uses `doWork()` to execute JaVers' Metadata verification, and it works perfectly.

So my questions are:

1. Does anyone perhaps know of a creative way to get the current combination of JaVers, JTA, and Spring Boot 3 to work together nicely?
2. If not, are there perhaps any plans to migrate the JaVers project to using the new `doWork()` and `doReturningWork()` methods rather than operating directly on the connection object?
3. Failing that, does it mean I need to effectively create my own implementations of any JaVers classes that execute SQL to use the above method?

Any suggestions would be greatly appreciated :)

**Comment by bartoszwalacik** [comment]
Hi @justin-schreuder ,
thanks for the good problem statement. Imo the way to go is to migrate JaVers to use the new doWork() and doReturningWork() methods. There are no plans on my side to do it though. Fell free to contribute the PR
